### PR TITLE
Implement SpanProcessor::shutdown()

### PR DESCRIPTION
**Pull request by michaelorr**
The diff isn't too big here, but there are several logical changes all at once... I'm leaning towards breaking this PR up into a few parts to make it more manageable to merge, but none of the individual moving parts would make sense without the bigger picture of what they enabled, so it may actually be hard to review in chunks without seeing the end state.

This change also:
- Implements `shutdown()` for the 2 Exporter implementations that we currently have
- Makes `Exporters` optional for `SpanProcessors`
- Changes the `BatchSpanProcessor` such that if either the configured batch size OR
  the configured delay is reached, then a batch is sent, both are no longer required
- Adds `SpanProcessor::forceFlush()`. The spec requires this and it aided
  the implementation of `BatchSpanProcessor::shutdown()`

I didn't see any logging throughout the code that exists. Some of the other libraries do some logging. I've never authored a php library so I wasn't sure if good solutions exist to enable us to log when (for example) the `SpanProcessor->$maxQueueSize` is reached.

Other thoughts are littered throughout the change.

### CODE REVIEW

File: sdk/Trace/BatchSpanProcessor.php [hunk 1]
```diff
@@ -11,7 +11,7 @@
 class BatchSpanProcessor implements SpanProcessor
 {
     /**
-     * @var Exporter
+     * @var Exporter|null
```
Comment (michaelorr) [inline sdk/Trace/BatchSpanProcessor.php:14]: > Each processor registered on TracerProvider is a start of pipeline that consist of span processor and optional exporter. SDK MUST allow to end each pipeline with individual exporter.

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#span-processor

File: sdk/Trace/BatchSpanProcessor.php [hunk 2]
```diff
@@ -77,11 +82,23 @@ public function onStart(API\Span $span): void
...
     public function onEnd(API\Span $span): void
     {
-        if (count($this->queue) < $this->maxQueueSize) {
-            $this->queue[] = $span;
+        if (null !== $this->exporter && $this->running) {
+            if (count($this->queue) < $this->maxQueueSize) {
+                $this->queue[] = $span;
+            }
+
+            if ($this->bufferReachedExportLimit() || $this->enoughTimeHasPassed()) {
```
Comment (michaelorr) [inline sdk/Trace/BatchSpanProcessor.php:90]: The spec is very vague on the interplay between `maxExportBatchSize` and `scheduledDelayMillis` but both the python and go SDK implementations appear to send a batch for whichever occurs first, not necessarily waiting for both conditions to be met. This change brings the php sdk in line with the other implementations.

File: sdk/Trace/TracerProvider.php [hunk 3]
```diff
@@ -43,6 +43,15 @@ class TracerProvider implements API\TracerProvider
...
         $this->spanProcessors = $spanProcessors;
+
+        register_shutdown_function([$this, 'shutdown']);
+    }
+
+    public function shutdown(): void
+    {
+        foreach ($this->spanProcessors as $spanProcessor) {
+            $spanProcessor->shutdown();
+        }
```
Comment (michaelorr) [inline sdk/Trace/TracerProvider.php:54]: This `TracerProvider::shutdown()` helper is only needed while the TracerProvider is managing the array of `$spanProcessors`... once https://github.com/open-telemetry/opentelemetry-php/pull/94 is merged, this helper can go away and the callback registration in the constructor can look like:
`register_shutdown_function([$this->spanProcessors, 'shutdown'])` since it will handle the "fan-out" to all the other downstream SpanProcessors

File: sdk/Trace/BatchSpanProcessor.php (hunk 2 above)
Comment (bobstrecansky) [reply sdk/Trace/BatchSpanProcessor.php:90]: I think that the `whichever occurs first` scenario seems to make sense to me.  We should send a batch when we run out of space or time :+1:

File: sdk/Trace/SimpleSpanProcessor.php [hunk 4]
```diff
@@ -23,24 +28,38 @@ public function __construct(Exporter $exporter)
...
 
     /**
      * @inheritDoc
      */
     public function onEnd(API\Span $span): void
     {
-        // @todo only spans with SampleFlag === true should be exported according to
-        // https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/sdk-tracing.md#sampling
-        $this->exporter->export([$span]);
+        if ($this->running) {
```
Comment (bobstrecansky) [inline sdk/Trace/SimpleSpanProcessor.php:38]: Should we address this todo in this PR or do you think it should be in a separate PR?

**Review by bobstrecansky — COMMENTED** [review]
I think this is a great start!  You mentioned to me in DM that you were going to add more tests.  Otherwise I very much like the direction of this PR.

**Comment by michaelorr** [comment]
This is no longer needed because https://github.com/open-telemetry/opentelemetry-php/pull/100
